wpilibj/src/main/java/edu/wpi/first/wpilibj/shuffleboard/ContainerHelper.java

```
@@ -0,0 +1,88 @@
+ */
+final class ContainerHelper {
+
+  private final ShuffleboardContainer m_container;
+  private final Set<String> m_usedTitles = new HashSet<>();
+  private final List<ShuffleboardComponent<?>> m_components = new ArrayList<>();
+  private final Map<String, ShuffleboardLayout> m_layouts = new LinkedHashMap<>(); // NOPMD
+
+  ContainerHelper(ShuffleboardContainer container) {
+    m_container = container;
+  }
+
+  List<ShuffleboardComponent<?>> getComponents() {
+    return m_components;
+  }
+
+  ShuffleboardLayout getLayout(String type, String title) {
+    if (!m_layouts.containsKey(title)) {
+      ShuffleboardLayout layout = new ShuffleboardLayout(m_container, type, title);
+      m_components.add(layout);
+      m_layouts.put(title, layout);
+    }
+    return m_layouts.get(title);
+  }
+
+  ComplexWidget add(String title, Sendable sendable) {
+    checkTitle(title);
+    ComplexWidget widget = new ComplexWidget(m_container, title, sendable);
+    m_components.add(widget);
+    return widget;
+  }
+
+  ComplexWidget add(Sendable sendable) throws IllegalArgumentException {
+    if (sendable.getName() == null || sendable.getName().isEmpty()) {
+      throw new IllegalArgumentException("Sendable must have a name");
+    }
+    return add(sendable.getName(), sendable);
+  }
+
+  SimpleWidget add(String title, Object data) {
+    Objects.requireNonNull(title, "Title cannot be null");
+    Objects.requireNonNull(data, "Data cannot be null");
+    checkTitle(title);
+    checkNtType(data);
+
+    SimpleWidget widget = new SimpleWidget(m_container, title);
+    m_components.add(widget);
+    widget.getEntry().forceSetValue(data);
+    return widget;
+  }
+
+  private static void checkNtType(Object data) {
+    if (!NetworkTableEntry.isValidDataType(data)) {
+      throw new IllegalArgumentException(
+          "Cannot add data of type " + data.getClass().getName() + " to Shuffleboard");
+    }
```

Do we really want to throw an exception here?

---

Yes. This API is intended to be used at startup in `robotInit`, so user code will crash at startup with an invalid Shuffleboard configuration.